docs/architecture/fsm-1.md

```diff
@@ -0,0 +1,78 @@
+# Connection Handshake FSM
+
+
+## Input Events:
+
+
+###### Chain events:
+
+- __OpenInit__
+- __OpenTry__
+- __OpenAck__
+- __OpenConfirm__
+- __NewBlock__
+- __UpdateClient__
+
+
+###### Query results:
+
+- __RelayerClientResponse__, parametrized result:
+    * if (y), then there is some new headers are available and the client on the opposite chain should be updated
+    * if (n), then the client on the opposite chain does not need to be updated
+
+
+## FSM states:
+
+1. __Z__ = uninitialized
+2. __Q[d]__ = querying (fetching) headers from destination chain
+3. __Q[s]__ = querying (fetching) headers from source chain
+4. __U[s]__ = updating IBC client on source chain with (newer) headers
+5. __U[d]__ = updating IBC client on destination chain with (newer) headers
+6. __W__ = waiting for new block event (next height) on the source chain
+7. __S__ = submitting the datagram + proofs to destination chain
+8. __X__ = terminated
+
+Proof-related states:
+
+- __P__ = creating proofs
+- __PCC__ = creating the client (or consensus) proof + connection state proof
+- __PC__ = creating the connection state proof
+
+
+## FSM transition table:
+
+
+| State|           Input        | Next State |  Output  |
+|:----:|:----------------------:|:----------:|:--------:|
+| Z    | OpenInit \| OpenTry    |   Q[d]     | Query destination chain headers via relayer client |
+| Z    | OpenAck |      W     | _ |
+| Z    | OpenConfirm |      X     | FSM Exit |
+| Q[d] | RelayerClientResponse(y)|   U[s]     | Submit UpdateClient on source chain |
+| Q[d] | RelayerClientResponse(n)|   W     | _ |
+| U[s] | UpdateClient           |   Q[s]     | Query source chain headers via relayer client |
+| W    | NewBlock               |   Q[s]     | Query source chain headers via relayer client |
+| Q[s] | RelayerClientResponse(y) |   U[d]     | Submit UpdateClient on destination chain |
+| Q[s] | RelayerClientResponse(n) |   P     | [internal] Create proofs |
+| U[d] | UpdateClient           |    P       | [internal] Create proofs |
+| P    | OpenInit \| OpenTry    |   PCC      | Query the IBC client state and the connection state on source chain |
+| P    | OpenAck     |    PC      | Query the connection state on source chain |
+| P    | OpenConfirm     |    X      | FSM Exit |
+| PC   | RelayerClientResponse(n) |    S      | Submit datagram + proofs to destination chain |
+| PCC  | RelayerClientResponse(n) |    S      | Submit datagram + proofs to destination chain |
+|  S   | NewBlock               |   X       | FSM Exit |
```

Comment:
Why exit on new block?

Reply:
This NewBlock event signals that the datagram and proof (submitted in state __S__) arrived and processes at the destination chain. Is there further processing we'd like to do as part of the FSM?